src/main/java/org/gridsuite/directory/server/DirectoryService.java

```diff
@@ -89,31 +89,46 @@ private static ElementAttributes toElementAttributes(DirectoryElementEntity enti
         return Mono.fromRunnable(() -> directoryElementRepository.updateElementAccessRights(UUID.fromString(directoryUuid), accessRightsAttributes.isPrivate()));
     }
 
-    public Mono<Void> deleteElement(String elementUuid, String userId) {
-        return Mono.fromRunnable(() -> deleteElementTree(elementUuid, userId));
+    public Mono<ElementAttributes> getElementInfos(String directoryUuid) {
+        return Mono.fromCallable(() -> directoryElementRepository.findById(UUID.fromString(directoryUuid)).map(DirectoryService::toElementAttributes).orElse(null));
     }
 
-    private void deleteElementTree(String elementUuid) {
-        directoryContentStream(elementUuid).forEach(e -> {
-            deleteElementTree(e.getElementUuid().toString());
-            if (e.getType().equals(ElementType.STUDY)) {
-                deleteStudy(e.getElementUuid()).subscribe();
+    public Mono<Void> deleteElement(String elementUuid, String userId) {
+        return getElementInfos(elementUuid).map(elementAttributes -> {
+            if (elementAttributes.getType().equals(ElementType.STUDY)) {
+                deleteStudy(elementAttributes.getElementUuid(), userId).subscribe();
+                directoryElementRepository.deleteById(UUID.fromString(elementAttributes.getElementUuid().toString()));
+            } else {
+                deleteElementTree(elementUuid, userId);
             }
-        });
+            return elementAttributes;
+        }).then();
+    }
+
+    private void deleteElementTree(String elementUuid, String userId) {
+        deleteSubElements(elementUuid, userId);
         directoryElementRepository.deleteById(UUID.fromString(elementUuid));
     }
 
-    public Mono<ElementAttributes> getElementInfos(String directoryUuid) {
-        return Mono.fromCallable(() -> directoryElementRepository.findById(UUID.fromString(directoryUuid)).map(DirectoryService::toElementAttributes).orElse(null));
+    private void deleteSubElements(String elementUuid, String userId) {
+        directoryContentStream(elementUuid, userId).forEach(e -> {
+            if (e.getType().equals(ElementType.STUDY)) {
```

Comment:
This part of code is similar to the one present in deleteElement method. Maybe it can be mutualized.

Reply:
Done

Reply:
In fact, the part of code in the "else" block can also be put in the mutualized method.